ssl/quic/quic_channel.c

```diff
@@ -113,6 +114,91 @@ static int gen_rand_conn_id(OSSL_LIB_CTX *libctx, size_t len, QUIC_CONN_ID *cid)
     return 1;
 }
 
+static unsigned long chan_reset_token_hash(
+        const QUIC_STATELESS_RESET_TOKEN *tok)
+{
+    unsigned long h;
+
+    assert(sizeof(h) <= sizeof(tok->token));
+    memcpy(&h, tok->token, sizeof(h));
+    return h;
+}
+
+static int chan_reset_token_cmp(const QUIC_STATELESS_RESET_TOKEN *t1,
+                                const QUIC_STATELESS_RESET_TOKEN *t2)
+{
+    return CRYPTO_memcmp(t1->token, t2->token, sizeof(t1->token));
+}
+
+static int reset_token_obfuscate(QUIC_STATELESS_RESET_TOKEN *out,
+                                 const unsigned char *in, size_t len)
+{
+    /*
+     * TODO(QUIC): update this to AES encrypt the token in ECB mode with a
+     * random (per channel) key.
+     */
+    memcpy(out, in, sizeof(*out));
+    return 1;
+}
+
+/*
+ * Remove the old token from the table of legal tokens and insert the new
+ * one in it's place.
+ */
+static int chan_update_reset_token(QUIC_CHANNEL *ch,
+                                   const unsigned char *new, size_t new_len)
+{
+    /*
+     * Remove any existing reset token, if this fails it means there wasn't
+     * one and this can safely be ignored.
+     */
+    (void)lh_QUIC_STATELESS_RESET_TOKEN_delete(ch->stateless_reset_tokens,
+                                               &ch->cur_remote_reset_token);
+    reset_token_obfuscate(&ch->cur_remote_reset_token, new, new_len);
+    lh_QUIC_STATELESS_RESET_TOKEN_insert(ch->stateless_reset_tokens,
+                                         &ch->cur_remote_reset_token);
+    return !lh_QUIC_STATELESS_RESET_TOKEN_error(ch->stateless_reset_tokens);
```

Comment:
I am not sure this is right. Although we retire the connection id immediately, it does not mean the token is really retired. IMO it can be counted as retired only when the peer acks the packet with the retire frame. However we do not need to check the ack - we can just save the current and the previous reset token as there cannot be more than 2 connection ids active - this is our active connection id limit.

Reply:
The standard is clear (5.1.2) that on receipt of a new ID frame, the old IDs must not be used anymore and must be retired before the new ID is added.  It also notes (10.3.1) that reset tokens associated with retired CIDs need not be remembered.

And confusion is introduced in 10.3 and 19.16 by saying the CID/token is retired via a retire frame so the peer will retain it until then and could potentially use it :disappointed: 


The big question for me is: how would we determine that the peer has received our ACK and we can remove the old reset token?  The only way to tell for sure would be when we don't ever receive a retransmission of the new ID frame.   I.e. we cannot.

Reply:
Let's assume this scenario - our active connection id limit is 2, the peer issues new connection id without retiring the previous one (it is allowed to do that). We won't use it but we won't retire it (as retiring it regardless of the peer asking for it being retired could trigger a loop with the peer constantly adding new connection ids). Then two stateless reset tokens are active and should be able to terminate the connection.

So at minimum we need to support 2 tokens.

Reply:
What you're asking for is the separation of the retirement and the new.  That I can do easily.

Reply:
Looks like we've got a complete implementation of stateless resets now.
Keeping two was the same as keeping any number.